### DISCUSSION TITLE
Integrating Custom Oracle Function for Field Decryption in SQLAlchemy ORM Model

### DISCUSSION THREAD

**Discussion by dtuecks** [sqlalchemy/sqlalchemy#11072] [Usage Questions]
We've recently started transitioning from raw cx_Oracle queries to SQLAlchemy 2.0 in ORM mode. Up to now we are using a simple ORM model to fetch data from an Oracle database like this:

```
class SOME_TABLE(Base):
    __tablename__ = "SOME_TABLE"
    USER_ID: Mapped[str] = mapped_column(String, primary_key=True)
    USER_ATTR1: Mapped[str]
    USER_ATTR2: Mapped[str]
```

Now, I've encountered a field that is encrypted using a custom Oracle function. To decrypt it, I need to call this function with a token as an argument. We have an existing Python method _getToken(key)_ that generates a valid token. The process in a non-ORM setup looks something like this:

```
mytoken = getToken(key)
SELECT USERNAME, USER_ATTR1, USER_ATTR2, ENCRYPT_IFC.decrypt_text(USER_ATTR3, mytoken) FROM SOME_TABLE
```

The _getToken(key)_ method should be called right before the query execution.

I'm trying to figure out how to translate this into an ORM model. Initially, I thought I could retrieve the token using an @orm.reconstructor function and then possibly use a @hybrid_property to execute my custom select. However, I've learned that @orm.reconstructor is invoked after the data has been loaded from the database, which doesn't fit my use case.

Does anyone have suggestions on how I can achieve this? Any advice or examples would be greatly appreciated.

- Daniel

**Comment by CaselIT** [comment]
Hi,

A couple of other notes: usually the class attribute use lowercase names, since sqlalchemy will thread lowercase names as "case insensitive" that in oracle means they will be upper-case. More info here https://docs.sqlalchemy.org/en/20/dialects/oracle.html#identifier-casing

Regarding your question, off the top of my head I would suggest you to use an hybrid property https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html:

```py
class SOME_TABLE(Base):
    __tablename__ = "SOME_TABLE"
    USER_ID: Mapped[str] = mapped_column(String, primary_key=True)
    USER_ATTR1: Mapped[str]
    USER_ATTR2: Mapped[str]

    _user_attr3: Mapped[str] = mapped_column('user_attr3')  # pass the name of the column in the table
   
    @hybrid_property
    def user_attr3(self):
      # this is in python
      return self._user_attr3
   
    @user_attr3.inplace.expression
    @classmethod
    def _user_attr3_expression(cls):
      # this is in the sql
      return sa.func.ENCRYPT_IFC.decrypt_text(cls._user_attr3, getToken(key))

```

there is also the possibility of defining setters

**Reply by dtuecks** [reply]
First of all, thank you very much for your input! It's greatly appreciated. Using an expression was one of the first things I tried (unsuccessfully). 

Based on your suggestion, I modified the code as follows:

```
class SOME_TABLE(Base):
    __tablename__ = "SOME_TABLE"
    _user_attr3: Mapped[str] = mapped_column('user_attr3', nullable=False, primary_key=True)  
    
    @hybrid_property
    def user_attr3(self):
        print("===== HYBRID =====")
        return self._user_attr3
    
    @user_attr3.inplace.expression
    @classmethod
    def _user_attr3_expression(cls):
        print("===== EXPRESSION =====")
        return sa.func.ENCRYPT_IFC.decrypt_text(cls._user_attr3, token.getToken(key))
```

I'm calling it like this:

```
from model import SOME_TABLE

result = session.query(SOME_TABLE).all()
print(result[0].user_attr3)
```

And here is the output:

```
2024-03-01 15:12:45,756 INFO sqlalchemy.engine.Engine select sys_context('userenv', 'current_schema') from dual
2024-03-01 15:12:45,756 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-03-01 15:12:45,775 INFO sqlalchemy.engine.Engine SELECT value FROM v$parameter WHERE name = 'compatible'
2024-03-01 15:12:45,775 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-03-01 15:12:45,802 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-01 15:12:45,805 INFO sqlalchemy.engine.Engine SELECT "SOME_TABLE".user_attr3 AS "SOME_TABLE_user_attr3" FROM "SOME_TABLE"
2024-03-01 15:12:45,805 INFO sqlalchemy.engine.Engine [generated in 0.00034s] {}
===== HYBRID =====
<encrypted value of user_attr3>
```

It appears that **_user_attr3_expression(cls)** is never called since I can't see "**===== EXPRESSION =====**" in the output. Additionally, I would expect the log line to include the SQL function call, like:

`2024-03-01 15:12:45,805 INFO sqlalchemy.engine.Engine SELECT ENCRYPT_IFC.decrypt_text(user_attr3, <value-of-token>) AS "SOME_TABLE_user_attr3" FROM "SOME_TABLE"`

What might I be missing?

**Reply by CaselIT** [reply]
It may be that's because you set it as primary key. Can you try not setting it as pk in the model?

**Reply by dtuecks** [reply]
I've tried, but it doesn't seem to make a difference. The model now looks like this (I added 'user_attr1' as PK, removed PK and nullable from _user_attr3):

```
class SOME_TABLE(Base):
    __tablename__ = "SOME_TABLE"
    user_attr1: Mapped[str] = mapped_column('user_attr1', nullable=False, primary_key=True)  
    _user_attr3: Mapped[str] = mapped_column('user_attr3')
    
    @hybrid_property
    def user_attr3(self):
        print("===== HYBRID =====")
        return self._user_attr3
    
    @user_attr3.inplace.expression
    @classmethod
    def _user_attr3_expression(cls):
        print("===== EXPRESSION =====")
        return sa.func.ENCRYPT_IFC.decrypt_text(cls._user_attr3, token.getToken(key))
```
        
Still, it only prints out "===== HYBRID =====" and not "===== EXPRESSION =====". Also, the string is not decrypted and the SQL debug output does not show any signs of ENCRYPT_IFC.decrypt_text.

btw: I'm using 

```
oracledb==2.0.1
SQLAlchemy==2.0.27
```

**Reply by CaselIT** [reply]
If you add to the select the user_attr3 property directly it does work but that's not what you are looking for.
Probably hybrid property it not the best option here, sorry.

Using a column property works but I don't think it allows setting new values:
```py
import sqlalchemy as sa
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase, column_property
from sqlalchemy.ext.hybrid import hybrid_property


class Base(DeclarativeBase): ...


def getToken(key):
    return key.upper()


key = "this is a key"


class SOME_TABLE(Base):
    __tablename__ = "SOME_TABLE"
    USER_ID: Mapped[str] = mapped_column(sa.String, primary_key=True)
    USER_ATTR1: Mapped[str]
    USER_ATTR2: Mapped[str]

    _user_attr3: Mapped[str] = mapped_column(
        "user_attr3"
    )  # pass the name of the column in the table

    user_attr3: Mapped[str] = column_property(
        sa.func.ENCRYPT_IFC.decrypt_text(_user_attr3, getToken(key))
    )


print(sa.select(SOME_TABLE))
```

prints
```sql
SELECT ENCRYPT_IFC.decrypt_text("SOME_TABLE".user_attr3, :decrypt_text_2) AS decrypt_text_1, "SOME_TABLE"."USER_ID", "SOME_TABLE"."USER_ATTR1", "SOME_TABLE"."USER_ATTR2", "SOME_TABLE".user_attr3
FROM "SOME_TABLE"
```
